lib/SPIRV/libSPIRV/SPIRVModule.cpp

```diff
@@ -2103,70 +2105,275 @@ static std::string to_string(VersionNumber Version) {
               to_string(VersionNumber::MinimumVersion) + " - " +
               to_string(VersionNumber::MaximumVersion))) {
-    M.setInvalid();
+    MI.setInvalid();
     return I;
   }
 
-  bool SPIRVVersionIsAllowed = M.isAllowedToUseVersion(MI.SPIRVVersion);
-  if (!M.getErrorLog().checkError(
+  bool SPIRVVersionIsAllowed = MI.isAllowedToUseVersion(MI.SPIRVVersion);
+  if (!ErrorLog.checkError(
           SPIRVVersionIsAllowed, SPIRVEC_InvalidModule,
           "incorrect SPIR-V version number " + to_string(MI.SPIRVVersion) +
               " - it conflicts with maximum allowed version which is set to " +
-              to_string(M.getMaximumAllowedSPIRVVersion()))) {
-    M.setInvalid();
+              to_string(MI.getMaximumAllowedSPIRVVersion()))) {
+    MI.setInvalid();
+    return I;
+  }
+
+  SPIRVWord Generator = ReadSPIRVWord(I);
+  if (!ErrorLog.checkError(!I.fail(), SPIRVEC_InvalidModule,
+                           "header parsing error")) {
+    MI.setInvalid();
     return I;
   }
 
-  SPIRVWord Generator = 0;
-  Decoder >> Generator;
   MI.GeneratorId = Generator >> 16;
   MI.GeneratorVer = Generator & 0xFFFF;
 
   // Bound for Id
-  Decoder >> MI.NextId;
+  MI.NextId = ReadSPIRVWord(I);
+  if (!ErrorLog.checkError(!I.fail(), SPIRVEC_InvalidModule,
+                           "header parsing error")) {
+    MI.setInvalid();
+    return I;
+  }
 
-  Decoder >> MI.InstSchema;
-  if (!M.getErrorLog().checkError(MI.InstSchema == SPIRVISCH_Default,
-                                  SPIRVEC_InvalidModule,
-                                  "unsupported instruction schema")) {
-    M.setInvalid();
+  MI.InstSchema = static_cast<SPIRVInstructionSchemaKind>(ReadSPIRVWord(I));
+  if (!ErrorLog.checkError(MI.InstSchema == SPIRVISCH_Default,
+                           SPIRVEC_InvalidModule,
+                           "unsupported instruction schema")) {
+    MI.setInvalid();
     return I;
   }
 
-  while (Decoder.getWordCountAndOpCode() && M.isModuleValid()) {
-    SPIRVEntry *Entry = Decoder.getEntry();
-    if (Entry != nullptr)
-      M.add(Entry);
+  SPIRVWord WordCount = 0;
+  Op OpCode = OpNop;
+  SPIRVEntry *Scope = nullptr;
+  while (true) {
+    WordCount = ReadSPIRVWord(I);
+    if (I.fail()) {
+      SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode FAIL 0 0\n");
+      break;
+    }
+    std::string RawOp;
+    I >> RawOp;
+    OpCode = getNameMap(OpCode).rmap(RawOp);
+    SPIRVDBG(spvdbgs() << "Read word: W = " << RawOp << " V = " << OpCode
+                       << '\n');
+    if (I.fail()) {
+      SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode FAIL 0 0\n");
+      break;
+    }
+    SPIRVDBG(spvdbgs() << "[SPIRVDecoder] getWordCountAndOpCode " << WordCount
+                       << " " << OpCodeNameMap::map(OpCode) << '\n');
+    if (!MI.isModuleValid()) {
+      break;
+    }
+
+    SPIRVEntry *Entry =
+        parseAndCreateSPIRVEntry(WordCount, OpCode, Scope, MI, I);
```

Comment:
We can just call this parseAndCreateEntry?

Reply:
The name of the function in this case have a precise meaning - `SPIRVEntry` which is translator internal type used for Entity/Object parsed by parser. It's part of current internal parser algorithm and not something which should be used by anyone outside parser scope. That's also why the symbol is not exported externally.